jmclient/jmclient/payjoin.py

```diff
@@ -0,0 +1,503 @@
+
+def send_payjoin(manager, accept_callback=None,
+                 info_callback=None, tls_whitelist=None):
+    """ Given a JMPayjoinManager object `manager`, initialised with the
+    payment request data from the server, use its wallet_service to construct
+    a payment transaction, with coins sourced from mixdepth `mixdepth`,
+    then wait for the server response, parse the PSBT, perform checks and complete sign.
+    The info and accept callbacks are to ask the user to confirm the creation of
+    the original payment transaction (None defaults to terminal/CLI processing),
+    and are as defined in `taker_utils.direct_send`.
+
+    If `tls_whitelist` is a list of bytestrings, they are treated as hostnames
+    for which tls certificate verification is ignored. Obviously this is ONLY for
+    testing.
+
+    Returns:
+    (True, None) in case of payment setup successful (response will be delivered
+     asynchronously) - the `manager` object can be inspected for more detail.
+    (False, errormsg) in case of failure.
+    """
+
+    # wallet should already be synced before calling here;
+    # we can create a standard payment, but have it returned as a PSBT.
+    assert isinstance(manager, JMPayjoinManager)
+    assert manager.wallet_service.synced
+    payment_psbt = direct_send(manager.wallet_service, manager.amount, manager.mixdepth,
+                             str(manager.destination), accept_callback=accept_callback,
+                             info_callback=info_callback,
+                             with_final_psbt=True)
+    if not payment_psbt:
+        return (False, "could not create non-payjoin payment")
+
+    manager.set_payment_tx_and_psbt(payment_psbt)
+    # TODO add delayed call to broadcast this after 1 minute
+
+    # Now we send the request to the server, with the encoded
+    # payment PSBT
+    if not tls_whitelist:
+        agent = Agent(reactor)
+    else:
+        agent = Agent(reactor,
+            contextFactory=WhitelistContextFactory(tls_whitelist))
+
+    body = BytesProducer(payment_psbt.to_base64().encode("utf-8"))
+    # TODO what to use as user agent?
+    d = agent.request(b"POST", manager.server.encode("utf-8"),
+        Headers({"Content-Type": ['text/plain']}),
+        bodyProducer=body)
+
+    d.addCallback(receive_payjoin_proposal_from_server, manager)
+    # note that the errback (here "noresponse") is *not* triggered
+    # by a server rejection (which is accompanied by a non-200
+    # status code returned), but by failure to communicate.
+    def noResponse(failure):
+        failure.trap(ResponseFailed)
+        log.error(failure.value.reasons[0].getTraceback())
+        reactor.stop()
+    d.addErrback(noResponse)
+    return (True, None)
+
+def fallback_nonpayjoin_broadcast(manager, err):
+    assert isinstance(manager, JMPayjoinManager)
+    log.warn("Payjoin did not succeed, falling back to non-payjoin payment.")
+    log.warn("Error message was: " + str(err))
```

Comment:
`err.decode('utf-8')` should be used here instead of `str(err)`, currently outputs `[WARNING]  Error message was: b'Service Unavailable'`.

Reply:
OK will do (is it technically possible it might crash if it doesn't decode? but i guess there is no such scenario in reality).

Reply:
Done in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/536/commits/13aab6117dd3885375d78088dec84f8fc2b1c7e7